ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiSnapshotWorldStateKeyValueStorage.java

```diff
@@ -67,12 +70,19 @@ public BonsaiUpdater updater() {
         trieLogStorage);
   }
 
+  @Override
+  public boolean isClosed() {
+    return isClosed.get();
+  }
+
   @Override
   public void close() throws Exception {
-    accountStorage.close();
-    codeStorage.close();
-    storageStorage.close();
-    trieBranchStorage.close();
+    if (!isClosed.getAndSet(true)) {
+      accountStorage.close();
+      codeStorage.close();
+      storageStorage.close();
+      trieBranchStorage.close();
+    }
```

Comment:
Since we are loading the cache async, I think we can still end up trying to read from a closed db.  I think a better pattern would be to prevent _actual_ closing the storage while there is still a consumer running an operation.  

Something along the lines of passing a consumer of the worldstate into a `boolean safeExecute(Consumer<KeyValueStorage> x)` function that will ensure the worldstate does not close while the lambda is running.  if the storage is closed, the function should return false to indicate the storage is closed.

We shouldn't block calls to `close()` on that function, just have close() mark the actual closing of the storages.  both `close` and `safeExecute` can call `private void doClose()`at the end of their function to close if both the keyvaluestorage has been marked "to_close" , and no outstanding safeExecute operations are running.